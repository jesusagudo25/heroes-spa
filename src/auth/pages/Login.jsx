import { useNavigate } from 'react-router-dom';

export const Login = () => {

  const navigate = useNavigate();

  const onLogin = () => {
    //navigate('/');
    navigate('/', { replace: true });
  }
  return (
    <div className="container mt-5">

      <h1>Login</h1>

      <hr />

      <div className="d-flex justify-content-center">
        <button className="btn btn-primary" onClick={onLogin}>
          Login
        </button>
      </div>
    </div>
  )
}
