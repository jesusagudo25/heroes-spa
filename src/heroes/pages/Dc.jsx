
import { HeroList } from '../components';

export const Dc = () => {
  return (
    <div className="container">
      
      <h1>DC Comics</h1>
      <hr />

      <HeroList publisher="DC Comics" />
              
    </div>
  )
}
