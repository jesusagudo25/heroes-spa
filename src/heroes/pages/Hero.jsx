import { useMemo } from "react";
import { useParams, Navigate, useNavigate } from "react-router-dom";
import { getHeroById } from "../helpers";

export const Hero = () => {

  const { heroId } = useParams();
  //Esta funcion se podria optimizar para que no se ejecute cada vez que se renderiza el componente, es decir. Si ya se cargo la informacion del heroe, no se deberia volver a llamar a la funcion getHeroById. Las opciones son usar useMemo o usar un custom hook, LocalStorage, etc.
  //const hero = getHeroById(heroId);
  const hero = useMemo(() => getHeroById(heroId), [heroId]);
  //Funcionamiento: Si el valor de heroId cambia, se vuelve a ejecutar la funcion getHeroById, si no cambiara, no se ejecutara la funcion getHeroById.

  const navigate = useNavigate();

  const onReturn = () => {
    if (hero.publisher === 'Marvel Comics') {
      navigate('/marvel');
    } else {
      navigate('/dc');
    }

    //Or navigate(-1);
  };

  if (!hero) {
    //not found - 404 (OR )
    return <Navigate to="/" />;
  }

  return (
    <div>
      <h1>Hero</h1>
      <hr />
      <div className="row">
        <div className="col-4">
          <img
            src={`/assets/imgs/heroes/${hero.id}.jpg`}
            alt={hero.superhero}
            className="img-thumbnail animate__animated animate__fadeInLeft"
          />
        </div>
        <div className="col-8">
          <h3>{hero.superhero}</h3>
          <ul className="list-group list-group-flush mb-3">
            <li className="list-group-item">
              <b>Alter ego:</b> {hero.alter_ego}
            </li>
            <li className="list-group-item">
              <b>Publisher:</b> {hero.publisher}
            </li>
            <li className="list-group-item">
              <b>First appearance:</b> {hero.first_appearance}
            </li>
          </ul>
          <h5>Characters</h5>
          <p>{hero.characters}</p>
          <button className="btn btn-outline-info" onClick={onReturn}>
            Return
          </button>
        </div>
      </div>
    </div>
  )
}
