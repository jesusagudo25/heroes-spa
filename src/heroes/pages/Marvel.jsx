import { HeroList } from '../components';

export const Marvel = () => {
  return (
    <div className="container">
      
      <h1>Marvel Comics</h1>
      <hr />

      <HeroList publisher="Marvel Comics" />
              
    </div>
  )
}
