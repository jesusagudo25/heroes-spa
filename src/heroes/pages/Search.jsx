import { HeroCard } from '../components'
import { useForm } from '../../hooks/'
import { useLocation, useNavigate } from 'react-router-dom'
import queryString from 'query-string'
import { getHeroesByName } from '../helpers'
import { useMemo } from 'react'

export const Search = () => {

  const navigate = useNavigate();
  const location = useLocation();

  const { q = '' } = queryString.parse(location.search);
  const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

  const showSearch = (q === '') ;
  const showNotFound = (q !== '' && heroesFiltered.length === 0);

  const { searchText, handleInputChange } = useForm({
    searchText: q
  });

  const onSearchSubmit = (e) => {
    e.preventDefault();
    /*     if (searchText.trim().length <= 1) return;
     */
    navigate(`?q=${searchText.trim()}`);
  }

  return (
    <>
      <h1>Search</h1>
      <hr />

      <div className="row">

        <div className="col-5">
          <h4>Search Form</h4>
          <hr />
          <form onSubmit={onSearchSubmit}>
            <input
              type="text"
              placeholder="Find your hero"
              className="form-control"
              autoComplete="off"
              name="searchText"
              value={searchText}
              onChange={handleInputChange}
            />
            <button
              type="submit"
              className="btn mt-3 btn-block btn-outline-primary"
            >
              Search...
            </button>

            <button
              type="reset"
              className="btn mt-3 btn-block btn-outline-danger"
              style={{ marginLeft: '10px' }}
            >
              Reset
            </button>
          </form>
        </div>

        <div className="col-7" aria-label="heroes-col" style={{ display: 'block' }}>
          <h4>Results</h4>
          <hr />

          {
              /*
                (q === '') ?
                  <div className="animate__animated animate__fadeIn">
                    <div className="alert alert-info">Search a hero...</div>
                  </div>
                : (heroesFiltered.length === 0) && <div className="animate__animated animate__fadeIn">
                  <div className="alert alert-danger">Not hero with <b>{q}</b></div>
                  </div> 
              */
          }

          {
            showSearch && <div className="animate__animated animate__fadeIn">
              <div className="alert alert-info">Search a hero...</div>
            </div>
          }

          {
            showNotFound && <div className="animate__animated animate__fadeIn">
              <div className="alert alert-danger">Not hero with <b>{q}</b></div>
            </div>
          }

          {/* Tambien se puede hacer con un operador ternario, para que el style se aplique solo si se cumple la condicion: display: (q === '') ? 'none' : 'block' */}

          <div className="d-flex gap-2 flex-column">
            {heroesFiltered.map(hero => (
              <HeroCard key={hero.id} hero={hero} />
            ))}
          </div>
        </div>

      </div>
    </>
  )
}
