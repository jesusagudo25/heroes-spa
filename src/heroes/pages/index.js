export * from './Marvel';
export * from './Dc';
export * from './Search';
export * from './Hero';