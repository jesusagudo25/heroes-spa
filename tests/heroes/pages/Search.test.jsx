import { render, screen } from "@testing-library/react";
import { Search } from "../../../src/heroes/pages/Search";
import { MemoryRouter } from "react-router-dom";

describe('Test in <Search />', () => { 
    
    test('should display correctly with default values', () => {

        //Primero: renderizar el componente con el MemoryRouter, ya que este componente usa el hook useLocation, que es un hook de react-router-dom... Ademas, se le pasa la propiedad initialEntries con un array que contiene la ruta inicial que se quiere probar, finalmente se renderiza el componente Search y se destructura el container. El container es el contenedor que contiene el componente renderizado.
        const {container} = render(
        <MemoryRouter initialEntries={['/search']}>
            <Search />
        </MemoryRouter>
        );

        //Segundo: se espera que el container sea igual al snapshot, es decir, que el componente se renderizo correctamente. El objetivo de esto es que si en un futuro se hacen cambios en el componente, se pueda comparar con el snapshot y ver si los cambios son correctos.
        expect(container).toMatchSnapshot();
        
    });

    test('should display Batman and the input with the value of the queryString', () => {

        //Primero: renderizar el componente con el MemoryRouter, ya que este componente usa el hook useLocation, que es un hook de react-router-dom... Ademas, se le pasa la propiedad initialEntries con un array que contiene la ruta inicial que se quiere probar, finalmente se renderiza el componente Search y se destructura el container. El container es el contenedor que contiene el componente renderizado.
        const {container} = render(
        <MemoryRouter initialEntries={['/search?q=Batman']}>
            <Search />
        </MemoryRouter>
        );

        //Segundo: se espera que el container sea igual al snapshot, es decir, que el componente se renderizo correctamente. El objetivo de esto es que si en un futuro se hacen cambios en el componente, se pueda comparar con el snapshot y ver si los cambios son correctos.
        //expect(container).toMatchSnapshot();

        //Tercero: se espera que el input tenga el valor de Batman. Se utiliza textbox porque es un input de tipo texto.
        expect(screen.getByRole('textbox').value).toBe('Batman');

        //Cuarto: se espera que la imagen tenga el alt de Batman.
        const img = screen.getByRole('img');
        expect(img.alt).toBe('Batman');

       //screen.debug();
        const heroesCol = screen.getByLabelText('heroes-col'); 
        expect(heroesCol.style.display).toBe('block');
        
    });
 })