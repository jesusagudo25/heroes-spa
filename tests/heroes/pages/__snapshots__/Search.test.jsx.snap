// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <Search /> should display correctly with default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          class="btn mt-3 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
        <button
          class="btn mt-3 btn-block btn-outline-danger"
          style="margin-left: 10px;"
          type="reset"
        >
          Reset
        </button>
      </form>
    </div>
    <div
      aria-label="heroes-col"
      class="col-7"
      style="display: block;"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="animate__animated animate__fadeIn"
      >
        <div
          class="alert alert-info"
        >
          Search a hero...
        </div>
      </div>
      <div
        class="d-flex gap-2 flex-column"
      />
    </div>
  </div>
</div>
`;
